chain/checkpointing/sub.go

```diff
@@ -0,0 +1,600 @@
+		f := frost.KeygenTaproot(id, ids, threshold)
+
+		handler, err := protocol.NewMultiHandler(f, []byte{1, 2, 3})
+		if err != nil {
+			panic(err)
+		}
+		c.LoopHandler(ctx, handler, n)
+		r, err := handler.Result()
+		if err != nil {
+			panic(err)
+		}
+		fmt.Println("Result :", r)
+
+		c.newconfig = r.(*keygen.TaprootConfig)
+
+		return true, nil
+	}
+
+	revertHandler := func(ctx context.Context, ts *types.TipSet) error {
+		return nil
+	}
+
+	match := func(oldTs, newTs *types.TipSet) (bool, events.StateChange, error) {
+		// verify we are synced
+		st, err := c.api.SyncState(ctx)
+		if err != nil {
+			panic(err)
+		}
+		//fmt.Println(st.ActiveSyncs)
+
+		if !c.synced {
+			// Are we synced ?
+			if len(st.ActiveSyncs) > 0 &&
+				st.ActiveSyncs[len(st.ActiveSyncs)-1].Height == newTs.Height() {
+
+				fmt.Println("We are synced")
+				// Yes then verify our checkpoint
+				ts, err := c.api.ChainGetTipSet(ctx, c.latestConfigCheckpoint)
+				if err != nil {
+					panic(err)
+				}
+				fmt.Println("We have a checkpoint up to height : ", ts.Height())
+				c.synced = true
+				c.height = ts.Height()
+			} else {
+				return false, nil, nil
+			}
+		}
+
+		newAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, newTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		oldAct, err := c.api.StateGetActor(ctx, mpower.PowerActorAddr, oldTs.Key())
+		if err != nil {
+			return false, nil, err
+		}
+
+		var oldSt, newSt mpower.State
+
+		bs := blockstore.NewAPIBlockstore(c.api)
+		cst := cbor.NewCborStore(bs)
+		if err := cst.Get(ctx, oldAct.Head, &oldSt); err != nil {
+			return false, nil, err
+		}
+		if err := cst.Get(ctx, newAct.Head, &newSt); err != nil {
+			return false, nil, err
+		}
+
+		// Init with the first checkpoint for the demo
+		// Should be done outside of it ?
+		if !c.init && c.config != nil {
+			ts, err := c.api.ChainGetTipSetByHeight(ctx, 0, oldTs.Key())
+			if err != nil {
+				panic(err)
+			}
+			data := ts.Cids()[0]
+			err = c.initiate(data.Bytes())
+			if err != nil {
+				panic(err)
+			}
+			return false, nil, nil
+		}
+
+		// Activate checkpointing every 30 blocks
+		fmt.Println("Height:", newTs.Height())
+		if newTs.Height()%30 == 0 && c.config != nil {
+			fmt.Println("Check point time")
+
+			// Initiation and config should be happening at start
+			if c.init {
+				data := oldTs.Key().Bytes()
+
+				var config string = hex.EncodeToString(data) + "\n"
+				for _, partyId := range c.orderParticipantsList() {
+					config += partyId + "\n"
+				}
+
+				hash, err := CreateConfig([]byte(config))
+				if err != nil {
+					panic(err)
+				}
+
+				// Push config to S3
+				err = StoreConfig(ctx, c.minioClient, c.cpconfig.MinioBucketName, hex.EncodeToString(hash))
+				if err != nil {
+					panic(err)
+				}
+
+				go c.CreateCheckpoint(ctx, data, hash)
+			}
+		}
+
+		// If Power Actors list has changed start DKG
+		// Changes detected so generate new key
+		if oldSt.MinerCount != newSt.MinerCount {
```

Comment:
we may only need to check that the size of symmetrical difference between the two sets is bigger than some threshold u.